### PR TITLE
Added S3 support for datasets, refactored asset loaders

### PR DESCRIPTION
**Pull request by aittalam**
## What's changing

- Added S3 support for datasets
- Refactored the asset loaders to (1) support S3 datasets and (2) be asset-specific. We now have one class for datasets, one for models/configs, one for tokenizers.

## How to test it
- Used pytest for backwards compatibility.
- Manually tested S3 dataset loading in the prometheus job (we might want to rely on localstack for testing)

### CODE REVIEW

File: src/lm_buddy/jobs/asset_loader.py
Comment (sfriedowitz) [inline src/lm_buddy/jobs/asset_loader.py:27]: After I removed the "artifact loader" interface a few PRs ago, these classes no longer carry state (per the deleted comment below). It might be more fitting to simply turn these into functions instead of classes.

The reason I could see a class being useful is if the asset loader had some external dependency (e.g. an S3 client from boto3) that is required to do its job. But right now, thats not the case.

File: src/lm_buddy/jobs/evaluation/lm_harness.py
Comment (sfriedowitz) [inline src/lm_buddy/jobs/evaluation/lm_harness.py:40]: Same thing w/ my comment above. These classes have no internal state, so a function seems a bit cleaner.

**Review by sfriedowitz — COMMENTED** [review]
I have one comment to consider regarding asset loader classes vs helper functions. It's your call, so if you don't want to make the changes I can approve after your response.

**Comment by aittalam** [comment]
> I have one comment to consider regarding asset loader classes vs helper functions. It's your call, so if you don't want to make the changes I can approve after your response.

Thanks Sean! As S3 support for models & co (basically anything which is not a dataset) has a dependency on an s3 client, I think it makes sense to keep this (even if I agree with you that the function approach looks cleaner!).

**Review by sfriedowitz — APPROVED** [review]
LGTM

**Comment by sfriedowitz** [comment]
@aittalam whoops! Add a version bump before merging please.

**Comment by aittalam** [comment]
> @aittalam whoops! Add a version bump before merging please.

Great catch! 🙏 :-)

**Pull request merged**
